### DISCUSSION TITLE
[A-L] (18 maggio 2024) Esercizio 2

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#96] [Esercitazioni A-L]
![immagine](https://github.com/user-attachments/assets/87e2b6ec-4cfb-481d-8c7c-990b7df9b362)

**Comment by cornflexxx** [comment]
Creo la matrice $T_{n\times C}$ dove $T[k,c] \rightarrow$ valore massimo ottenibile con un sottoinsieme di oggetti da $a_1,...,a_k$ e con peso minore di $c$

$$T[k,c] = \left\\{ \begin{array}{cl}
0 & c = 0\lor k = 0 \\
T[k-1,c]& w_k \gt c\\
\max(T[k-1,c],T[k-1,c-w_k]+v_k & w_k\le c
\end{array} \right.$$

$T[n,C]$ è la soluzione.

**Comment by FeddyLix17** [comment]
Del [Knapsack problem](https://en.wikipedia.org/wiki/Knapsack_problem) esistono diverse [varianti](https://en.wikipedia.org/wiki/List_of_knapsack_problems) (tuttavia l'esercizio ci chiede di risolvere il problema nella sua forma base 0/1, e successivamente una variante di quest'ultimo).

Seguendo un approccio di programmazione dinamica (in questo caso tabulare), creiamo una matrice $\Large n+1 \times C+1$ per tenere traccia del massimo valore raggiungibile segliendo i primi $\Large k \in \lbrace 0, 1, \dots, n\rbrace$  elementi di A[] per ogni peso $\Large c \in \lbrace0, 1, \dots, C\rbrace$.

Per la stessa definizione, il valore massimo raggiungibile scegliendo sempre 0 elementi o avendo capienza sempre uguale a 0 sarà 0

<div align="center">

| $\Huge _k\setminus^c$ | 0 | 1 | $\dots$ | C |
| :-: | :-: | :-: | :-: | :-: |
| 0 | 0 | 0 | 0 | 0 |
| 1 | 0 | $\dots$ | $\dots$ | $\dots$ |
| $\dots$ | 0 | $\dots$ | $\dots$ | $\dots$ |
| n | 0 | $\dots$ | $\dots$ | $\dots$ |

</div>

Ogni $\Large T[k, C]$ contterrà il massimo valore raggiungibile scegliendo k elementi (e di conseguenza $\Large T[n, C]$ contterrà la soluzione al primo quesito.

Ad esempio, data una Capacità dello zaino C = 10, e 4 oggetti A, B, D, E riportati come (valore, peso), dunque

- A = (4, 3)

- B = (5, 4)

- D = (6, 5)

- E = (2, 3)

il massimo valore raggiungibile scegliendo solo un elemento (l'elemento A) sarà dato da

<div align="center">

$$
\Large
\begin{equation}
    T[1,c]=
    \begin{cases}
      T[k-1,c] = 0 & w_k > c \\
       w_k = 3 & w_k \leq c
    \end{cases}
  \end{equation}
$$

| $\Huge _k\setminus^c$ | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 0 | 3 | 3 | 3 | 3 | 3 | 3 | 3 | 3 |
| 2 | 0 | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ |
| 3 | 0 | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ |
| 4 | 0 | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ |

</div>

Già dalla scelta di almeno 2 oggetti, si arriva alla generalizzazione della formula, ovvero

$$
\Large
\begin{equation}
    T[k,c]=
    \begin{cases}
      \quad \quad \quad \quad \quad \quad T[k-1,c] & w_k > c \\
       max(T[k-1,c],\ T[k-1,c - w_k] + v_k) & w_k \leq c
    \end{cases}
  \end{equation}
$$

scegliendo il valore massimo raggiungibile (per la stessa capienza c) tra

- quello ottenuto precedentemente non scegliendo l'oggetto $\Large A_k$

- e quello ottenuto scegliendolo senza sforare il peso c corrente

<div align="center">

| $\Huge _k\setminus^c$ | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 0 | 4 | 4 | 4 | 4 | 4 | 4 | 4 | 4 |
| 2 | 0 | 0 | 0 | 4 | 5 | 5 | 5 | 9 | 9 | 9 | 9 |
| 3 | 0 | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ |
| 4 | 0 | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ | $\dots$ |

</div>

Proseguendo con il riempimento della tabella, si arriverà alla seguente situazione

<div align="center">

| $\Huge _k\setminus^c$ | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 0 | 4 | 4 | 4 | 4 | 4 | 4 | 4 | 4 |
| 2 | 0 | 0 | 0 | 4 | 5 | 5 | 5 | 9 | 9 | 9 | 9 |
| 3 | 0 | 0 | 0 | 4 | 5 | 6 | 6 | 9 | 10 | 11 | 11 |
| 4 | 0 | 0 | 0 | 4 | 5 | 6 | 6 | 9 | 10 | 11 | $\textnormal{\textcolor{green}{11}}$ |

</div>

Una possibile implementazione

```python

def knapsack(A, C):
    
# A viene fornito come lista di elementi nella forma
# i-esimo elemento(valore, peso)

# C viene fornito come intero > 0

n = len(A)

# inizializzazione tabella n + 1 X C + 1
T = []
for i in range(n + 1):
    riga = []
    for j in range(C + 1):
        riga.append(0)
    T.append(riga)

# implementazione effettiva dell'algoritmo
for k in range(1, n + 1):
    for c in range(C + 1):
        if A[k][1] > c:
            T[k][c] = T[k - 1][c]
        else:
            T[k][c] = max(T[k - 1][c], T[k - 1][c - A[k][1]] + A[k][0])

return T[n][C]

```
Entrambe le coppie di cicli for presentato tutte operazioni elementari al loro interno, rispettando cosi il costo di $\Large O(n \cdot c)$.

Per ritornare anche gli elementi che compongo una possibile soluzione basterà ripercorrere all'[indietro](https://en.wikipedia.org/wiki/Backtracking) la tabella T secondo il seguente criterio

- partendo da $\Large T[k][c]$, con k = n e c = C, trovare che il valore presente a  $\Large T[k - 1][C]$ sia diverso da esso significherebbe dire che l'elemento k faccia parte dell'insieme $\Large U \subseteq A$

- a quel punto, la ricerca proseguirebbe su $\Large T[k-1][c - w_k]$ poichè sarebbero da considerare solo i gli altri sottoinsiemi nei quali l'elemento k sicuramente non sia presente

la lista ottenuta conterrà gli elementi di una possibile soluzione

```python
def knapsack(A, C):

n = len(A)

# inizializzazione tabella n + 1 X C + 1
T = []
for i in range(n + 1):
    riga = []
    for j in range(C + 1):
        riga.append(0)
    T.append(riga)

# implementazione effettiva dell'algoritmo
for k in range(1, n + 1):
    for c in range(C + 1):
        if A[k][1] > c:
            T[k][c] = T[k - 1][c]
        else:
            T[k][c] = max(T[k - 1][c], T[k - 1][c - A[k][1]] + A[k][0])

# back-tracking sulla tabella T
capienzaCorrente = C
U = []

for k in range(n, 0, -1):
    if A[k][capienzaCorrente] != A[k - 1][capienzaCorrente]:
        U.append(k)
        capienzaCorrente -= A[k - 1][1]

return T[n][C], U
```

il nuovo ciclo for aggiunto presenta tutte operazioni elementari al suo interno e complessità $\Large O(n)$, mantenendo invariato il costo dell'intero algoritmo.

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/3ace3d54-cd5c-4cf2-a28f-0871c2ea84ce)
